commit 4f20a1dc5242fb4bb8763e0451df898fa48e740c
Author: Martin Sebor <msebor@redhat.com>
Date:   Tue Jan 25 17:39:36 2022 -0700

    stdlib: Avoid -Wuse-after-free in __add_to_environ [BZ #26779]
    
    Reviewed-by: Carlos O'Donell <carlos@redhat.com>

diff --git a/stdlib/setenv.c b/stdlib/setenv.c
index 893f081af6b5a21b..14fff422a2193864 100644
--- a/stdlib/setenv.c
+++ b/stdlib/setenv.c
@@ -150,7 +150,9 @@ __add_to_environ (const char *name, const char *value, const char *combined,
     {
       char **new_environ;
 
-      /* We allocated this space; we can extend it.  */
+      /* We allocated this space; we can extend it.  Avoid using the raw
+	 reallocated pointer to avoid GCC -Wuse-after-free.  */
+      uintptr_t ip_last_environ = (uintptr_t)last_environ;
       new_environ = (char **) realloc (last_environ,
 				       (size + 2) * sizeof (char *));
       if (new_environ == NULL)
@@ -159,7 +161,7 @@ __add_to_environ (const char *name, const char *value, const char *combined,
 	  return -1;
 	}
 
-      if (__environ != last_environ)
+      if ((uintptr_t)__environ != ip_last_environ)
 	memcpy ((char *) new_environ, (char *) __environ,
 		size * sizeof (char *));
 
